libraries/GCS_MAVLink/GCS.h

```diff
@@ -437,11 +431,77 @@ class GCS_MAVLINK
     char _perf_packet_name[16];
     char _perf_update_name[16];
 
-    // deferred message handling.  We size the deferred_message
-    // ringbuffer so we can defer every message type
-    enum ap_message deferred_messages[MSG_LAST];
-    uint8_t next_deferred_message;
-    uint8_t num_deferred_messages;
+    // outbound ("deferred message") queue.
+
+    // this is a simple array of deferred_message_t.  We keep a
+    // pointer to the next message that is to be sent (by offset,
+    // next_deferred_message_to_send); we need only check its
+    // send_not_before_ms entry to determine if we have work to do.
+    // In the event the message is sent, its send_not_before_ms is
+    // updated, and we linearly scan the array to find the next
+    // message that should be sent.  We may immediately send that
+    // message, or may find it isn't due to be sent yet, or that we
+    // have hit some sort of time limit.  An interval of 0 indicates
+    // that this message should not be rescheduled once sent.  An
+    // send_not_before_ms of zero indicates this message should not be
+    // sent.
+
+    // structure holding information about a scheduled message.
+    typedef struct {
+        enum ap_message id;
```

Comment:
Yes.  And that means some thing's rates will always be linked.  The alternative - another layer of indirection between mavlink IDs and `ap_message` IDs - was ghastly (in hindsight).

If it turns out to be a real problem then the `ap_message` can be split into multiple `ap_message`s

Reply:
It might be OK to have some messages linked although I wasn't expecting that - that's a group decision that a dev call should take.

Just to be clear: my issue isn't the use of `ap_message` per-se, but the fact that some messages we send can't have the interval set.